File: tsasdk-java/tsa-sdk-git/src/main/java/cn/signdoc/client/TSAClient.java
package cn.signdoc.client;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.math.BigInteger;
import java.net.URL;
import java.net.URLConnection;
import java.security.Security;
import java.text.SimpleDateFormat;

import lombok.extern.slf4j.Slf4j;
import org.bouncycastle.asn1.ASN1ObjectIdentifier;
import org.bouncycastle.asn1.cmp.PKIFailureInfo;
import org.bouncycastle.cms.CMSSignedData;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.tsp.*;
import org.bouncycastle.util.encoders.Base64;

/**
 * @author Rick Diao
 */
@Slf4j
public class TSAClient {

	private String tsaURL;
	private String tsaUsername;
	private String tsaPassword;
	private int ConnectTimeout = 8000;

	public void setTsaURL(String tsaURL) {
		this.tsaURL = tsaURL;
	}
	public void setTsaUsername(String tsaUsername) {
		this.tsaUsername = tsaUsername;
	}
	public void setTsaPassword(String tsaPassword) {
		this.tsaPassword = tsaPassword;
	}
	public void setConnectTimeout(int connectTimeout) {
		ConnectTimeout = connectTimeout;
	}

	/**
	 * get time
	 * @param tsaToken
	 * @return String
	 * @throws Exception
	 */
	public String getTime(byte[] tsaToken) throws Exception {
		try {
			Security.addProvider(new BouncyCastleProvider());
			CMSSignedData csd = new CMSSignedData(tsaToken);
			TimeStampToken timeStampToken = new TimeStampToken(csd);
			SimpleDateFormat dateFm = new SimpleDateFormat(
					"yyyy-MM-dd HH:mm:ss");
			String tsaStrTime = dateFm.format(timeStampToken.getTimeStampInfo()
					.getGenTime());
			return tsaStrTime;
		} catch (Exception e) {
			throw e;
		}
	}

	/**
	 * get tsa
	 * @param hashCode
	 * @return byte[]
	 * @throws Exception
	 */
	public byte[] getTsaAndResp(byte[] hashCode, ASN1ObjectIdentifier hashOID)
			throws Exception {
		byte[] tsa = null;
		try {
			if(this.tsaURL == null && this.tsaUsername == null && this.tsaPassword == null){
				throw new Exception("timestamp account is null");
			}else{
				Security.addProvider(new BouncyCastleProvider());
				TimeStampRequestGenerator tsqGenerator = new TimeStampRequestGenerator();
				tsqGenerator.setCertReq(true);
				BigInteger nonce = BigInteger.valueOf(System.currentTimeMillis());
				TimeStampRequest request = tsqGenerator.generate(hashOID, hashCode, nonce);
				byte[] requestBytes = request.getEncoded();
				byte[] respBytes = getTSAResponse(requestBytes);

				TimeStampResponse response = new TimeStampResponse(respBytes);
				PKIFailureInfo failure = response.getFailInfo();
				int value = (failure == null) ? 0 : failure.intValue();
				if (value != 0)
					throw new Exception("Invalid TSA '" + tsaURL + "' response, code " + value);
				TimeStampToken tsaToken = response.getTimeStampToken();
				if (tsaToken == null)
					throw new Exception("TSA '" + tsaURL + "' failed to return time stamp token");
				tsa = tsaToken.getEncoded();
			}
			return tsa;
		} catch (Exception e) {
			throw e;
		} catch (Throwable t) {
			throw new Exception("Failed to get TSA response from '" + tsaURL
					+ "'", t);
		}
	}

	protected byte[] getTSAResponse(byte[] requestBytes) throws Exception {
		URL url = new URL(tsaURL);
		URLConnection tsaConnection = (URLConnection) url.openConnection();
		tsaConnection.setDoInput(true);
		tsaConnection.setDoOutput(true);
		tsaConnection.setUseCaches(false);
		tsaConnection.setConnectTimeout(this.ConnectTimeout);
		tsaConnection.setRequestProperty("Content-Type",
				"application/timestamp-query");
		tsaConnection.setRequestProperty("Content-Transfer-Encoding", "binary");
		if ((tsaUsername != null) && !tsaUsername.equals("")) {
			String userPassword = tsaUsername + ":" + tsaPassword;
			tsaConnection.setRequestProperty("Authorization", "Basic "
					+ new String(new sun.misc.BASE64Encoder()
							.encode(userPassword.getBytes())));
		}
		OutputStream out = tsaConnection.getOutputStream();
		out.write(requestBytes);
		out.close();
		InputStream inp = tsaConnection.getInputStream();
		ByteArrayOutputStream baos = new ByteArrayOutputStream();
		byte[] buffer = new byte[1024];
		int bytesRead = 0;
		while ((bytesRead = inp.read(buffer, 0, buffer.length)) >= 0) {
			baos.write(buffer, 0, bytesRead);
		}
		byte[] respBytes = baos.toByteArray();
		String encoding = tsaConnection.getContentEncoding();
		
		if (encoding != null && encoding.equalsIgnoreCase("base64")) {
			respBytes = Base64.decode(new String(respBytes));
		}
		return respBytes;
	}

}

File: tsasdk-csharp/Tsp/TSAClient.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Tsp;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Tsp;
using Org.BouncyCastle.Crypto.Digests;
using SignDoc.TsaSdk.Crypto;

namespace SignDoc.TsaSdk.Tsp
{
    /// <summary>
    /// RFC 3161 compliant TSA client.
    /// Supports SHA-256, SHA-384, SHA-512, and SM3 digest algorithms.
    /// </summary>
    public class TSAClient
    {
        private readonly string _tsaUrl;
        private readonly string _username;
        private readonly string _password;
        private readonly int _timeout;
        private readonly DigestAlgorithm _defaultAlgorithm;

        /// <summary>
        /// Create a new TSA client.
        /// </summary>
        /// <param name="tsaUrl">URL of the TSA server.</param>
        /// <param name="username">Username for Basic Auth (optional).</param>
        /// <param name="password">Password for Basic Auth (optional).</param>
        /// <param name="timeout">HTTP request timeout in milliseconds (default 8000).</param>
        /// <param name="defaultAlgorithm">Default digest algorithm (default SHA-256).</param>
        public TSAClient(
            string tsaUrl,
            string username = "",
            string password = "",
            int timeout = 8000,
            DigestAlgorithm defaultAlgorithm = DigestAlgorithm.SHA256)
        {
            _tsaUrl = tsaUrl ?? throw new ArgumentNullException(nameof(tsaUrl));
            _username = username ?? "";
            _password = password ?? "";
            _timeout = timeout;
            _defaultAlgorithm = defaultAlgorithm;
        }

        /// <summary>
        /// Request a timestamp token for the given data.
        /// </summary>
        /// <param name="data">The data to timestamp.</param>
        /// <param name="algorithm">Override the default digest algorithm.</param>
        /// <returns>The raw timestamp token bytes.</returns>
        public byte[] TimestampData(byte[] data, DigestAlgorithm? algorithm = null)
        {
            var algo = algorithm ?? _defaultAlgorithm;
            byte[] hash = HashUtil.ComputeHash(data, algo);
            return TimestampHash(hash, algo);
        }

        /// <summary>
        /// Request a timestamp token for a pre-computed hash.
        /// </summary>
        /// <param name="hashBytes">The pre-computed hash digest.</param>
        /// <param name="algorithm">The algorithm used to compute the hash.</param>
        /// <returns>The raw timestamp token bytes.</returns>
        public byte[] TimestampHash(byte[] hashBytes, DigestAlgorithm? algorithm = null)
        {
            var algo = algorithm ?? _defaultAlgorithm;
            string digestOid = OidConstants.GetDigestOid(algo);

            // Build TimeStampRequest
            var requestGenerator = new TimeStampRequestGenerator();
            requestGenerator.SetCertReq(true);

            var nonce = BigInteger.ValueOf(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
            var asn1Oid = new DerObjectIdentifier(digestOid);
            var request = requestGenerator.Generate(asn1Oid, hashBytes, nonce);
            byte[] requestBytes = request.GetEncoded();

            // Send to TSA
            byte[] responseBytes = GetTSAResponse(requestBytes);

            // Parse response
            var response = new TimeStampResponse(responseBytes);

            if (response.GetFailInfo() != null)
            {
                throw new Exception($"TSA returned failure: {response.GetFailInfo()}");
            }

            TimeStampToken token = response.TimeStampToken;
            if (token == null)
            {
                throw new Exception("TSA failed to return time stamp token");
            }

            return token.GetEncoded();
        }

        /// <summary>
        /// Extract the timestamp time from a token.
        /// </summary>
        /// <param name="tokenBytes">The raw timestamp token bytes.</param>
        /// <returns>The generation time as a string.</returns>
        public string GetTime(byte[] tokenBytes)
        {
            var tsToken = new TimeStampToken(new Org.BouncyCastle.Cms.CmsSignedData(tokenBytes));
            return tsToken.TimeStampInfo.GenTime.ToString("yyyy-MM-dd HH:mm:ss");
        }

        private byte[] GetTSAResponse(byte[] requestBytes)
        {
            var request = (HttpWebRequest)WebRequest.Create(_tsaUrl);
            request.Method = "POST";
            request.ContentType = "application/timestamp-query";
            request.Timeout = _timeout;
            request.ContentLength = requestBytes.Length;

            // Basic Auth
            if (!string.IsNullOrEmpty(_username))
            {
                string auth = Convert.ToBase64String(
                    Encoding.UTF8.GetBytes($"{_username}:{_password}"));
                request.Headers.Add("Authorization", $"Basic {auth}");
            }

            using (var stream = request.GetRequestStream())
            {
                stream.Write(requestBytes, 0, requestBytes.Length);
            }

            using (var response = (HttpWebResponse)request.GetResponse())
            {
                if (response.StatusCode != HttpStatusCode.OK)
                {
                    throw new Exception($"TSA returned HTTP {(int)response.StatusCode}");
                }

                using (var ms = new MemoryStream())
                {
                    response.GetResponseStream().CopyTo(ms);
                    byte[] respBytes = ms.ToArray();

                    string encoding = response.Headers["Content-Transfer-Encoding"];
                    if (encoding != null && encoding.Equals("base64", StringComparison.OrdinalIgnoreCase))
                    {
                        respBytes = Convert.FromBase64String(Encoding.ASCII.GetString(respBytes));
                    }

                    return respBytes;
                }
            }
        }
    }
}
